src/systemd.c

```diff
@@ -0,0 +1,272 @@
+/* SPDX-License-Identifier: LGPL-2.1-only */
+/**
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ * Author: Tom Hromatka <tom.hromatka@oracle.com>
+ * Author: Silvia Chapa <silvia.chapa@oracle.com>
+ */
+
+#include <libcgroup-internal.h>
+#include <systemd/sd-bus.h>
+#include <libcgroup.h>
+#include <unistd.h>
+#include <assert.h>
+#include <errno.h>
+
+#define ONE_SEC 1000000
+
+static const char * const modes[] = {
+	"fail",			/* CGROUP_SYSTEMD_MODE_FAIL */
+	"replace",		/* CGROUP_SYSTEMD_MODE_REPLACE */
+	"isolate",		/* CGROUP_SYSTEMD_MODE_ISOLATE */
+	"ignore-dependencies",	/* CGROUP_SYSTEMD_MODE_IGNORE_DEPS */
+	"ignore-requirements",	/* CGROUP_SYSTEMD_MODE_IGNORE_REQS */
+};
+static_assert((sizeof(modes) / sizeof(modes[0])) == CGROUP_SYSTEMD_MODE_CNT,
+	      "modes[] array must be same length as CGROUP_SYSTEMD_MODE_CNT");
+
+static const char * const sender = "org.freedesktop.systemd1";
+static const char * const path = "/org/freedesktop/systemd1";
+static const char * const interface = "org.freedesktop.systemd1.Manager";
+
+int cgroup_set_default_scope_opts(struct cgroup_systemd_scope_opts * const opts)
+{
+	if (!opts)
+		return ECGINVAL;
+
+	opts->delegated = 1;
+	opts->mode = CGROUP_SYSTEMD_MODE_FAIL;
+	opts->pid = 0;
+
+	return 0;
+}
+
+/*
+ * Returns time elapsed in usec
+ *
+ * Inspired-by: https://github.com/cockpit-project/cockpit/blob/main/src/tls/socket-io.c#L39
+ */
+static int64_t elapsed_time(const struct timespec * const start, const struct timespec * const end)
+{
+	int64_t elapsed = (end->tv_sec - start->tv_sec) * 1000000 +
+			  (end->tv_nsec - start->tv_nsec) / 1000;
+
+	assert(elapsed >= 0);
+
+	return elapsed;
+}
+
+static int job_removed_callback(sd_bus_message *message, void *user_data, sd_bus_error *error)
```

Comment:
Can this function be converted into `void`, because we return only success?

Reply:
I don't think so.  systemd is expecting a return value from it.

Reply:
https://github.com/systemd/systemd/blob/main/src/systemd/sd-bus.h#L111

Reply:
Agreed, sorry for the noise here.